feat: Axios 인스턴스 및 인터셉터 개선 (#26)

anneedesign-ui opened the pull request:
## 기능 요약 

- 공통 axios 인스턴스를 `src/api/apiClient.ts`로 분리하고,
  응답 인터셉터에서 401 기본 처리를 추가했습니다.
- 댓글 도메인 API 호출을 `src/comment/services.ts`의 `commentService`로 분리해,
  React Query 훅에서는 서비스 레이어만 사용하도록 하는 기반을 마련했습니다.

## 변경 내용

- `src/api/apiClient.ts`
  - `axiosInstance` 생성 (baseURL, withCredentials 설정)
  - 응답 인터셉터에서 401 상태 코드 공통 처리
  - `apiClient.get/post/put/delete` 래퍼 추가 (`response.data`만 반환)

- `src/comment/services.ts`
  - `commentService.getComments/createComment/updateComment/deleteComment` 구현
  - axios 직접 호출 제거, `apiClient`만 사용하도록 정리

## 고민한 부분

- apiClient에서 `config` 객체 전체를 받을지, `url, data, config` 시그니처로 갈지 비교했고
  호출부 가독성을 위해 `url, data, config` 방식으로 선택했습니다.
- 서비스 레이어에서 응답을 그대로 반환하고,
  에러/UX 처리는 후속 단계(Queries 레이어)에서 맡기도록 역할을 나눴습니다.

## 리뷰 요청 포인트

- `apiClient → commentService → React Query` 로 이어지는 레이어 분리가
  이 프로젝트 규모에서 유지보수/확장성 측면에 적절한지 보고 싶습니다.
- apiClient에서 `response.data`만 반환하는 현재 설계가 괜찮은지,
  아니면 응답 전체를 넘기고 도메인 서비스에서 가공하는 패턴을 선호하시는지도 의견 부탁드립니다.

## 관련 파일

- `src/api/apiClient.ts`
- `src/comment/services.ts`

anneedesign-ui committed:
feat: Axios 인스턴스 및 인터셉터 개선 (#26)

redfoot135 reviewed (commented):
**리뷰 요청 포인트**  
apiClient → commentService → React Query 로 이어지는 레이어 분리가 이 프로젝트 규모에서 유지보수/확장성 측면에 적절한지 보고 싶습니다.  
apiClient에서 response.data만 반환하는 현재 설계가 괜찮은지, 아니면 응답 전체를 넘기고 도메인 서비스에서 가공하는 패턴을 선호하시는지도 의견 부탁드립니다.

**시도·고민한 점 (Optional)**  
서비스 레이어에서 응답을 그대로 반환할지, `{ data, error }` 형태로 한 번 더 감쌀지 고민했는데, 우선은 React Query의 기본 에러 흐름과 충돌이 적은 “data만 반환” 형태로 선택했습니다.

**리뷰 받고 싶은 포인트 (Review Focus)**  
“상태 관리를 전역으로 관리하고 있는데, 구조가 적절한지 확인 부탁드립니다.”  
댓글 생성/수정 API가 실제로는 `Comment` 객체 전체를 주는지, `{ detail: string }`만 주는지에 따라 서비스 반환 타입을 `Comment`로 둘지, 별도의 `CommentResponse` 타입을 둘지 고민이 있습니다.

# 주강사 답변

## 1. API 레이어 분리

> apiClient -> commentService -> React Query -> page

레이어 분리는 적절합니다. 각자 담당해야하는 역할이 명확하게 분리되고 있습니다.  
하지만 조금 더 구조를 명확하게 하자면

> apiClient -> commentService -> store -> page

reactQuery도 하나의 wrapper로 추상화하여서 store에서 사용하는 것이 좋을겁니다.

## 2. apiClient에서 response.data만 반환하는 설계

현재 사용되는 정도로는 body만 반환하는 것도 충분히 괜찮습니다.  
이대로 사용하시다가 더 많은 데이터가 추가로 필요한 경우가 생긴다면 그때 apiClient를 확장하시면 됩니다.

확장할 때 return값 자체를 변경해버리면 기존 사용하던 함수들을 다 변경해야하는 번거로움이 생기죠. 그럴 때는 기존 return 값을 유지하면서 추가적인 데이터를 반환하는 방식을 사용하는 것도 좋은 방법이 될 것 같습니다.

```ts
import { AxiosRequestConfig, AxiosResponse } from 'axios';

// 오버로드 선언
export function get<T>(
  url: string,
  config?: Omit<AxiosRequestConfig, 'url'>
): Promise<T>;
export function get<T>(
  url: string,
  config: Omit<AxiosRequestConfig, 'url'>,
  options: { raw: true }
): Promise<AxiosResponse<T>>;

export async function get<T>(
  url: string,
  config?: Omit<AxiosRequestConfig, 'url'>,
  options?: { raw: boolean }
): Promise<T | AxiosResponse<T>> {
  const response = await axiosInstance.get<T>(url, config);
  return options?.raw ? response : response.data;
}
```

이런 느낌으로 옵셔널하게 raw 옵션을 줄 때 response자체를 리턴해주는 형식으로 기존 사용하던 코드에 영향을 주지 않고 사용할 수 있습니다.

## 3. 상태 전역관리
코멘트 관련해서 전역 상태관리 코드를 확인할 수가 없는데 만약 제가 잘못 체크한거면 디스코드 주 강사 질문방에 다시 올려주시면 좋을 것 같습니다.

## 4. 댓글 생성/수정 API return Type
저거는 고민할 일이 아닌게  
백엔드 응답 타입이 확정이 되어야 프런트에서도 확정해서 작업 완료를 할 수 있는 부분이 맞습니다.


사전에 정해진 api응답값이 있으면 그대로 쓰시고요. 기다려야하는 상황이면 임시로 지정하고 추후 확정되는 타입을 사용하시면 됩니다.

anneedesign-ui committed:
refactor: abstract React Query into comment store layer

- React Query 로직을 useCommentStore로 캡슐화
- page에서는 store만 사용하도록 추상화
- queryKey, mutation 로직을 store 내부로 숨김
- 라이브러리 교체 시에도 page 코드 수정 불필요

anneedesign-ui commented:
피드백 반영 완료했습니다! 🙏

### 변경사항
- React Query를 `useCommentStore`로 추상화했습니다
- page에서는 store의 인터페이스만 사용하도록 개선했습니다
- 라이브러리 교체 시에도 page 코드 수정 없이 store만 변경하면 되도록 설계했습니다

### 구조
apiClient → commentService → useCommentStore → page

재검토 부탁드립니다!

@G-jjunny
